### ISSUE TITLE
Exception: #<NoMethodError: undefined method `getc' for nil:NilClass> and other fun gateway issues

### ISSUE THREAD

**Issue by oct2pus** [meew0/discordrb#581]
for what its worth i've already moved on to a Golang rewrite of my bot which does not encounter this issue, but running the original ruby version of my [bot](https://github.com/oct2pus/jadebot/tree/Original-Ruby) on a bot account with around 500 guilds or more should reproduce this issue (which is a bit of an unreasonable request i admit). Might be useful information for anyone else who runs into this issue.

---

my connection continues to reconnect in an infinite loop, I seem to get these two error messages over and over. 

```
[ERROR : heartbeat @ 2018-10-28 15:27:22.650] The websocket connection has closed: nil
[ERROR : websocket @ 2018-10-28 15:27:22.650] An error occurred in the main websocket loop!
[ERROR : websocket @ 2018-10-28 15:27:22.651] Exception: #<NoMethodError: undefined method `getc' for nil:NilClass>
[ERROR : websocket @ 2018-10-28 15:27:22.657] /home/octo/legacy/jadebot/.vendor/bundle/ruby/2.3.0/bundler/gems/discordrb-a416edf5c8be/lib/discordrb/gateway.rb:593:in `block in websocket_loop'
[ERROR : websocket @ 2018-10-28 15:27:22.658] /home/octo/legacy/jadebot/.vendor/bundle/ruby/2.3.0/bundler/gems/discordrb-a416edf5c8be/lib/discordrb/gateway.rb:593:in `synchronize'
[ERROR : websocket @ 2018-10-28 15:27:22.658] /home/octo/legacy/jadebot/.vendor/bundle/ruby/2.3.0/bundler/gems/discordrb-a416edf5c8be/lib/discordrb/gateway.rb:593:in `websocket_loop'
[ERROR : websocket @ 2018-10-28 15:27:22.658] /home/octo/legacy/jadebot/.vendor/bundle/ruby/2.3.0/bundler/gems/discordrb-a416edf5c8be/lib/discordrb/gateway.rb:569:in `connect'
[ERROR : websocket @ 2018-10-28 15:27:22.658] /home/octo/legacy/jadebot/.vendor/bundle/ruby/2.3.0/bundler/gems/discordrb-a416edf5c8be/lib/discordrb/gateway.rb:463:in `block in connect_loop'
[ERROR : websocket @ 2018-10-28 15:27:22.658] /home/octo/legacy/jadebot/.vendor/bundle/ruby/2.3.0/bundler/gems/discordrb-a416edf5c8be/lib/discordrb/gateway.rb:462:in `loop'
[ERROR : websocket @ 2018-10-28 15:27:22.659] /home/octo/legacy/jadebot/.vendor/bundle/ruby/2.3.0/bundler/gems/discordrb-a416edf5c8be/lib/discordrb/gateway.rb:462:in `connect_loop'
[ERROR : websocket @ 2018-10-28 15:27:22.659] /home/octo/legacy/jadebot/.vendor/bundle/ruby/2.3.0/bundler/gems/discordrb-a416edf5c8be/lib/discordrb/gateway.rb:168:in `block in run_async'
[INFO : websocket @ 2018-10-28 15:27:22.659] Instant reconnection flag was set - reconnecting right away
```

```
[WARN : heartbeat @ 2018-10-28 15:31:33.193] Last heartbeat was not acked, so this is a zombie connection! Reconnecting
[ERROR : heartbeat @ 2018-10-28 15:31:33.193] The websocket connection has closed: nil
[INFO : websocket @ 2018-10-28 15:31:33.194] Instant reconnection flag was set - reconnecting right away
```

I have no clue how your gateway code works, but it *looks* like the problem is in heartbeat.rb and gateway.rb

**Comment by connorshea** [comment]
I can get it to show up semi-consistently on my [ouija-bot](https://github.com/connorshea/ouija-bot) deployed to Heroku. 

```
2018-11-11T18:45:06.858903+00:00 app[worker.1]: [INFO : websocket @ 2018-11-11 18:45:06.858] Instant reconnection flag was set - reconnecting right away
2018-11-11T18:45:11.603706+00:00 app[worker.1]: [INFO : websocket @ 2018-11-11 18:45:11.603] Discord using gateway protocol version: 6, requested: 6
2018-11-11T18:46:30.358539+00:00 app[worker.1]: [WARN : heartbeat @ 2018-11-11 18:46:30.358] Last heartbeat was not acked, so this is a zombie connection! Reconnecting
2018-11-11T18:46:30.358798+00:00 app[worker.1]: [ERROR : heartbeat @ 2018-11-11 18:46:30.358] The websocket connection has closed: nil
2018-11-11T18:46:30.358883+00:00 app[worker.1]: [ERROR : websocket @ 2018-11-11 18:46:30.358] An error occurred in the main websocket loop!
2018-11-11T18:46:30.358929+00:00 app[worker.1]: [ERROR : websocket @ 2018-11-11 18:46:30.358] Exception: #<NoMethodError: undefined method `getc' for nil:NilClass>
2018-11-11T18:46:30.358976+00:00 app[worker.1]: [ERROR : websocket @ 2018-11-11 18:46:30.358] /app/vendor/bundle/ruby/2.4.0/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:593:in `block in websocket_loop'
2018-11-11T18:46:30.359009+00:00 app[worker.1]: [ERROR : websocket @ 2018-11-11 18:46:30.358] /app/vendor/bundle/ruby/2.4.0/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:593:in `synchronize'
2018-11-11T18:46:30.359040+00:00 app[worker.1]: [ERROR : websocket @ 2018-11-11 18:46:30.359] /app/vendor/bundle/ruby/2.4.0/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:593:in `websocket_loop'
2018-11-11T18:46:30.359072+00:00 app[worker.1]: [ERROR : websocket @ 2018-11-11 18:46:30.359] /app/vendor/bundle/ruby/2.4.0/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:569:in `connect'
2018-11-11T18:46:30.359103+00:00 app[worker.1]: [ERROR : websocket @ 2018-11-11 18:46:30.359] /app/vendor/bundle/ruby/2.4.0/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:463:in `block in connect_loop'
2018-11-11T18:46:30.359134+00:00 app[worker.1]: [ERROR : websocket @ 2018-11-11 18:46:30.359] /app/vendor/bundle/ruby/2.4.0/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:462:in `loop'
2018-11-11T18:46:30.359166+00:00 app[worker.1]: [ERROR : websocket @ 2018-11-11 18:46:30.359] /app/vendor/bundle/ruby/2.4.0/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:462:in `connect_loop'
2018-11-11T18:46:30.359202+00:00 app[worker.1]: [ERROR : websocket @ 2018-11-11 18:46:30.359] /app/vendor/bundle/ruby/2.4.0/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:168:in `block in run_async'
2018-11-11T18:46:30.359247+00:00 app[worker.1]: [INFO : websocket @ 2018-11-11 18:46:30.359] Instant reconnection flag was set - reconnecting right away
```

It ends up in a loop doing that, sometimes instead looping like this without an explicit error:

```
2018-11-11T18:54:51.572548+00:00 app[worker.1]: [ERROR : heartbeat @ 2018-11-11 18:54:51.572] The websocket connection has closed: nil
2018-11-11T18:54:51.572637+00:00 app[worker.1]: [INFO : websocket @ 2018-11-11 18:54:51.572] Instant reconnection flag was set - reconnecting right away
2018-11-11T18:54:54.944807+00:00 app[worker.1]: [INFO : websocket @ 2018-11-11 18:54:54.944] Discord using gateway protocol version: 6, requested: 6
2018-11-11T18:56:15.075106+00:00 app[worker.1]: [WARN : heartbeat @ 2018-11-11 18:56:15.074] Last heartbeat was not acked, so this is a zombie connection! Reconnecting
2018-11-11T18:56:15.075150+00:00 app[worker.1]: [ERROR : heartbeat @ 2018-11-11 18:56:15.075] The websocket connection has closed: nil
2018-11-11T18:56:15.078342+00:00 app[worker.1]: [INFO : websocket @ 2018-11-11 18:56:15.078] Instant reconnection flag was set - reconnecting right away
2018-11-11T18:56:18.525321+00:00 app[worker.1]: [INFO : websocket @ 2018-11-11 18:56:18.525] Discord using gateway protocol version: 6, requested: 6
2018-11-11T18:57:38.576330+00:00 app[worker.1]: [WARN : heartbeat @ 2018-11-11 18:57:38.576] Last heartbeat was not acked, so this is a zombie connection! Reconnecting
2018-11-11T18:57:38.676576+00:00 app[worker.1]: [ERROR : heartbeat @ 2018-11-11 18:57:38.676] The websocket connection has closed: nil
2018-11-11T18:57:38.676651+00:00 app[worker.1]: [INFO : websocket @ 2018-11-11 18:57:38.676] Instant reconnection flag was set - reconnecting right away
2018-11-11T18:57:45.229555+00:00 app[worker.1]: [INFO : websocket @ 2018-11-11 18:57:45.229] Discord using gateway protocol version: 6, requested: 6
2018-11-11T18:59:02.181497+00:00 app[worker.1]: [WARN : heartbeat @ 2018-11-11 18:59:02.181] Last heartbeat was not acked, so this is a zombie connection! Reconnecting
```

**Comment by connorshea** [comment]
I tried "bisecting" the gem by testing with a bunch of different versions in my bot, but got pretty inconsistent results: https://gist.github.com/connorshea/cd9e8f9dbb1c19078e37756cc23c1b6f

tl;dr bisecting is pretty much useless in this case, and probably won't help us find the source of this issue. I don't see any obvious patterns here.

**Comment by egee-irl** [comment]
Tossing my hat in there - this bug affects my bot too. When it happens, my bot takes _seconds_ to respond to commands when normally he responds in less than a second.

**Comment by i-hardy** [comment]
I recently updated my [woolf](https://github.com/i-hardy/woolf/) bot to the rubygems release (had been locked to an old commit for a while), and immediately started seeing this error. The bot's on 291 servers and this seems to be happening a couple of times an hour.

```
[ERROR : websocket @ 2019-04-01 00:55:22.546] An error occurred in the main websocket loop!
[ERROR : websocket @ 2019-04-01 00:55:22.546] Exception: #<Errno::ECONNRESET: Connection reset by peer>
[ERROR : websocket @ 2019-04-01 00:55:22.546] /usr/local/lib/ruby/2.6.0/openssl/buffering.rb:57:in `sysread'
[ERROR : websocket @ 2019-04-01 00:55:22.546] /usr/local/lib/ruby/2.6.0/openssl/buffering.rb:57:in `fill_rbuff'
[ERROR : websocket @ 2019-04-01 00:55:22.546] /usr/local/lib/ruby/2.6.0/openssl/buffering.rb:98:in `read'
[ERROR : websocket @ 2019-04-01 00:55:22.546] /usr/local/lib/ruby/2.6.0/openssl/buffering.rb:262:in `getc'
[ERROR : websocket @ 2019-04-01 00:55:22.546] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:593:in `block in websocket_loop'
[ERROR : websocket @ 2019-04-01 00:55:22.546] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:593:in `synchronize'
[ERROR : websocket @ 2019-04-01 00:55:22.546] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:593:in `websocket_loop'
[ERROR : websocket @ 2019-04-01 00:55:22.546] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:569:in `connect'
[ERROR : websocket @ 2019-04-01 00:55:22.546] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:463:in `block in connect_loop'
[ERROR : websocket @ 2019-04-01 00:55:22.546] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:462:in `loop'
[ERROR : websocket @ 2019-04-01 00:55:22.546] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:462:in `connect_loop'
[ERROR : websocket @ 2019-04-01 00:55:22.546] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:168:in `block in run_async'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] The websocket connection has closed: nil
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] The websocket connection has closed due to an error!
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] Exception: #<Errno::EPIPE: Broken pipe>
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/lib/ruby/2.6.0/openssl/buffering.rb:322:in `syswrite'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/lib/ruby/2.6.0/openssl/buffering.rb:322:in `do_write'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/lib/ruby/2.6.0/openssl/buffering.rb:339:in `block in write'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/lib/ruby/2.6.0/openssl/buffering.rb:338:in `each'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/lib/ruby/2.6.0/openssl/buffering.rb:338:in `inject'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/lib/ruby/2.6.0/openssl/buffering.rb:338:in `write'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:804:in `send'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:414:in `send_packet'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:264:in `send_heartbeat'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:257:in `heartbeat'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:445:in `block (2 levels) in setup_heartbeats'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:438:in `loop'
[ERROR : heartbeat @ 2019-04-01 00:55:52.268] /usr/local/bundle/gems/discordrb-3.3.0/lib/discordrb/gateway.rb:438:in `block in setup_heartbeats'
```

**Comment by swarley** [comment]
Could you try setting `compress_mode: :large` in your bot initializer? This
is a change in 3.3.0 that is known to have effects on larger bots.

**Comment by localhostdotdev** [comment]
Can confirm this happens even in my small bot with 3-5 servers. I will try `compress_mode :large` to see if it fixes it.

**Comment by z64** [comment]
These connection problems are unrelated to this issue's topic

**Comment by egee-irl** [comment]
I've also been seeing this error for quite some time. I've been able to work around it by running my bot as a daemon (or Docker container) and setting up a cron job to restart it every hour. The errors that show up in my logs look like this:

```
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] An error occurred in the main websocket loop!
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] Exception: #<Errno::ECONNRESET: Connection reset by peer>
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] /usr/lib/ruby/2.5.0/openssl/buffering.rb:57:in `sysread'
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] /usr/lib/ruby/2.5.0/openssl/buffering.rb:57:in `fill_rbuff'
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] /usr/lib/ruby/2.5.0/openssl/buffering.rb:98:in `read'
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] /usr/lib/ruby/2.5.0/openssl/buffering.rb:262:in `getc'
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] /var/lib/gems/2.5.0/gems/discordrb-3.2.1/lib/discordrb/gateway.rb:544:in `block in websocket_loop'
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] /var/lib/gems/2.5.0/gems/discordrb-3.2.1/lib/discordrb/gateway.rb:544:in `synchronize'
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] /var/lib/gems/2.5.0/gems/discordrb-3.2.1/lib/discordrb/gateway.rb:544:in `websocket_loop'
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] /var/lib/gems/2.5.0/gems/discordrb-3.2.1/lib/discordrb/gateway.rb:525:in `connect'
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] /var/lib/gems/2.5.0/gems/discordrb-3.2.1/lib/discordrb/gateway.rb:432:in `block in connect_loop'
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] /var/lib/gems/2.5.0/gems/discordrb-3.2.1/lib/discordrb/gateway.rb:431:in `loop'
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] /var/lib/gems/2.5.0/gems/discordrb-3.2.1/lib/discordrb/gateway.rb:431:in `connect_loop'
Apr 02 10:25:22 localhost rake[20679]: [ERROR : websocket @ 2019-04-02 10:25:22.502] /var/lib/gems/2.5.0/gems/discordrb-3.2.1/lib/discordrb/gateway.rb:161:in `block in run_async'
Apr 02 10:25:40 localhost rake[20679]: [ERROR : heartbeat @ 2019-04-02 10:25:40.360] The websocket connection has closed: nil
Apr 02 10:25:40 localhost rake[20679]: [ERROR : heartbeat @ 2019-04-02 10:25:40.360] The websocket connection has closed due to an error!
Apr 02 10:25:40 localhost rake[20679]: [ERROR : heartbeat @ 2019-04-02 10:25:40.361] Exception: #<Errno::EPIPE: Broken pipe>
```

Considering this has been a known issue for so long, and there isn't much interest from the maintainers in fixing it, it might behoove the users of this gem to fork it.

At any rate, setting up a cron job to restart the bot at regular intervals is a decent-enough work around for this bug until it is fixed.

**Comment by Daniel-Worrall** [comment]
If you're able to identify reproducible steps for this bug, it'd help immensely in figuring out a solution for it. Also, please do fork the repository and don't forget to pull request your ideas. We love it when people want to contribute!

**Comment by z64** [comment]
**First - and this is for everyone in this thread - please read the issue topic!** This issue is about some bug in our internal logic causing a particular instance variable to become unset in concurrent code. This is *not* about broken pipes (or similar "normal" networking issues), or "The websocket connection has closed: nil" - that line simply means there was no information associated with the close frame. **The exceptions you are seeing are entirely unrelated, this is just a bug in some Ruby code.**

---

@egee-irl I can assure you we are *very* interested in fixing *any* issue with the library. The problem is that a lot of us cannot reproduce certain networking issues reliably *or at all*, and those who can lack the ability to debug the problem effectively (which may very well includes us maintainers), or users do not cooperate - a lot of our users are beginners to programming as a whole. Trust me, if I knew what was wrong and had enough information, I would happily pour (more) hours into fixing it. But, it is unfortunately not that simple. :disappointed: 

If you can provide additional network level debugging, or have other insight into the issues you're experiencing, please feel free to share them in the Discord channel where we can try to more closely work with you to resolve it, or another issue. As mentioned in the other comments, certain problems can occur because of poor performance with hardware, network, or just ruby in general. Steps like @swarley mentioned of disabling zlib-stream compression, if you haven't, may or may not help. You can see #600 as well on that topic - myself and the community did tons of research together to pinpoint that issue.

Again, please feel free to continue this discussion, preferably in Discord until more information can be obtained, or at least some other issue - but not here please :slightly_smiling_face:
